src/irmin-pack/inode.ml

```diff
@@ -59,6 +61,84 @@ struct
     module Metadata = Node.Metadata
   end
 
+  module Step =
+    Irmin.Hash.Typed
+      (H)
+      (struct
+        type t = T.step
+
+        let t = T.step_t
+      end)
+
+  module Child_ordering : Child_ordering with type step := T.step = struct
+    open T
+
+    type key = bytes
+
+    let log_entry = int_of_float (log (float Conf.entries) /. log 2.)
+
+    let () =
+      assert (log_entry >= 1);
+      (* NOTE: the [`Hash_bits] mode is restricted to inodes with at most 1024
+         entries in order to simplify the implementation (see below). *)
+      assert ((not (Conf.inode_child_order = `Hash_bits)) || log_entry <= 10);
+      assert (Conf.entries = int_of_float (2. ** float log_entry))
+
+    let key =
+      match Conf.inode_child_order with
+      | `Hash_bits ->
+          fun s -> Bytes.unsafe_of_string (hash_to_bin_string (Step.hash s))
+      | `Seeded_hash | `Custom _ ->
+          fun s -> Bytes.unsafe_of_string (step_to_bin_string s)
+
+    (* Assume [k = cryto_hash(step)] (see {!key}) and [Conf.entry] can
+       can represented with [n] bits. Then, [hash_bits ~depth k] is
+       the [n]-bits integer [i] with the following binary representation:
+
+         [k(n*depth) ... k(n*depth+n-1)]
+
+       When [n] is not a power of 2, [hash_bits] needs to handle
+       unaligned reads properly. *)
+    let hash_bits ~depth k =
+      assert (Bytes.length k = Step.hash_size);
+      (* We require that the child indices have at most 10 bits to ensure that
+         they span no more than 2 bytes of the step hash. The 3 byte case (with
+         [1 + 8 + 1]) does not happen for 10-bit indices because 10 is even, but
+         [2 + 8 + 1] would occur with 11-byte indices (e.g. when [depth=2]). *)
+      assert (log_entry <= 10);
```

Comment:
you are asserting this twice, is this necessary?

Reply:
No; removed.